src/activity_item_handling.cpp

```diff
@@ -1156,12 +1157,597 @@ void activity_on_turn_blueprint_move( player_activity &, player &p )
         p.assign_activity( act_multiple_construction );
         return;
     }
+}
+
+static bool is_dig_tool_nearby( player &p )
+{
+    const tripoint abspos = g->m.getabs( p.pos() );
+    const zone_manager &mgr = zone_manager::get_manager();
+    const std::unordered_set<tripoint> &src_set = mgr.get_near(
+                zone_type_id( "LOOT_TOOLS" ), abspos );
+    const zone_type_id no_pickup( "NO_NPC_PICKUP" );
+    bool found_required = false;
+    const std::vector<tripoint> &src_sorted = get_sorted_tiles_by_distance( abspos, src_set );
+    for( const tripoint &pabs : src_sorted ) {
+        tripoint p_loc = g->m.getlocal( pabs );
+        // dont pickup from this spot, even if its got what we need.
+        if( g->check_zone( no_pickup, p_loc ) ) {
+            continue;
+        }
+        // otherwise find the right tool
+        const auto items_at = g->m.i_at( p_loc );
+        for( const item &elem : items_at ) {
+            if( elem.has_quality( quality_id( "DIG" ), 1 ) ) {
+                found_required = true;
+                break;
+            }
+        }
+        if( found_required ) {
+            break;
+        }
+    }
+    return found_required;
+}
+
+static bool required_seeds_available( player *p, const std::string &seed_type )
+{
+    zone_manager &mgr = zone_manager::get_manager();
+
+    const tripoint abspos = g->m.getabs( p->pos() );
+    const zone_type_id no_pickup( "NO_NPC_PICKUP" );
 
-    // If we got here without restarting the activity, it means we're done.
-    if( p.is_npc() ) {
-        npc *guy = dynamic_cast<npc *>( &p );
-        guy->current_activity = "";
-        guy->revert_after_activity();
+    const std::unordered_set<tripoint> &src_set = mgr.get_near(
+                zone_type_id( "LOOT_SEEDS" ), abspos );
+    bool found_required = false;
+    std::vector<item *> seed_inv = p->items_with( [seed_type]( const item & itm ) {
+        return itm.typeId() == itype_id( seed_type );
+    } );
+    // we arent interested in comparing numbers to get the exact amounts to fetch
+    // we just check here incase itd be pointless to even start.
+    if( !seed_inv.empty() ) {
+        return true;
+    }
+
+    const std::vector<tripoint> &src_sorted = get_sorted_tiles_by_distance( abspos, src_set );
+    for( const tripoint &pabs : src_sorted ) {
+        tripoint p_loc = g->m.getlocal( pabs );
+        // dont pickup from this spot, even if its got what we need.
+        if( g->check_zone( no_pickup, p_loc ) ) {
+            continue;
+        }
+        // otherwise find the right seed
+        const auto items_at = g->m.i_at( p_loc );
+        for( const item &elem : items_at ) {
+            if( elem.typeId() == itype_id( seed_type ) ) {
+                found_required = true;
```

Comment:
Same thing with early return as in the function above.

Btw. why not extract the predicate and make a generic function that looks up whether an item with given predicate exists within a given zone, instead of repeating code.

Reply:
A predicate would be a lambda-type thing right? ( Apologies - I really am very new to c++ )

Reply:
fixed the early return, will de-duplicate and have a common function to extract locations from a predicate in a seperate PR ( when I know how to do them ).